### ISSUE TITLE
interceptArrayMutation disallows undefined array value

### ISSUE THREAD

**Issue by sisp** [xaviergonz/mobx-keystone#161]
The `interceptArrayMutation` function throws an error in development mode when an array element is `undefined`:

https://github.com/xaviergonz/mobx-keystone/blob/7fa164b1fdfd1af1dc06033c48faa1aa5cad90d2/packages/lib/src/tweaker/tweakArray.ts#L275-L283

and

https://github.com/xaviergonz/mobx-keystone/blob/7fa164b1fdfd1af1dc06033c48faa1aa5cad90d2/packages/lib/src/tweaker/tweakArray.ts#L319-L321

Using `undefined` as an array value is not wrong in general, although it cannot be JSON-serialized using the default serializer of course. I'd like to use `undefined` and have extended the JSON serializer and deserializer to deal with it. Would you be open to making it configurable whether `undefined` is allowed or not?

**Comment by xaviergonz** [comment]
The constraint is exactly for the reason you mention: compatibility with JSON serialization (getSnapshot / fromSnapshot / applySnapshot, etc).
If this were not enforced then getSnapshot would need to transform undefined to null and fromSnapshot transform null to undefined for elements inside the array, etc. which could be a source of confusion for a relative rare case.

One (naive) option is to use a computed view
```ts
@computed
get arrayWithUndefined() {
  return this.arrayWithNull.map(e => e === null ? undefined : e)
}
```
but that has the con that it would be slow / recomputed for any change inside the array.

What kind of configuration do you have in mind?

**Comment by sisp** [comment]
> If this were not enforced then getSnapshot would need to transform undefined to null and fromSnapshot transform null to undefined for elements inside the array

But these functions don't seralize to or deserialize from JSON themselves. It just happens that JSON (de)serialization is often done in this context.

Perhaps it's bad design on my side, but I distinguish between `undefined` and `null` in my data model, so both values are valid and have different semantics in my array. That's why I stumbled over the error.

> One (naive) option is to use a computed view
>
> ```
> @computed
> get arrayWithUndefined() {
>   return this.arrayWithNull.map(e => e === null ? undefined : e)
> }
> ```
>
> but that has the con that it would be slow / recomputed for any change inside the array.

Yes, I think the real problem is just JSON (de)serialization. I have extended the standard JSON serializer to turn `undefined` into the string `"<MY_UNIQUE_PREFIX>:undefined"` and the deserializer accordingly. Not sure if there's a better way of doing it, but I found this to be a working and pragmatic solution.

> What kind of configuration do you have in mind?

I was thinking of something like this (inspired by MobX' `configure` function): https://github.com/sisp/mobx-keystone/pull/1
